[posix] WS-Discovery: host name detection and lookup

mglae opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Implement host name detection via WS-Discovery for posix platforms and
resolve the names in DNSNameCache.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Only IP addresses are shown with #19971.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

On Linux X86_64 desktop against Windows, FritzBox and wsdd2 WS-Discovery implementation.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Servers can be selected by name, not only by IP address in local network.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

wsnipex commented on xbmc/network/DNSNameCache.cpp:
```diff
@@ -106,6 +112,13 @@ bool CDNSNameCache::Lookup(const std::string& strHostName, std::string& strIpAdd
+  if (WSInstance.GetCached(strHostName, strIpAddress))
+    return true;
+#endif
+
```
This will override cache entries added via <hosts> in advanced settings, so it should be moved further down

mglae replied:
Corrected.

mglae commented:
Test suite failure in TestWebServer.CanGetJsonRpcApiDescriptionWithHttpGet was related to this PR. Fixed.

fuzzard commented:
Short comment about this. It doesnt work on osx at all. Failure with nmblookup means it wont work on ios/tvos, and maybe not even android.

```
2021-09-28 13:21:12.194419+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.194 T:9393595   DEBUG <general>: CWSDiscoveryListenerUDP::DispatchCommand - Command sent
2021-09-28 13:21:12.195082+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.195 T:9393595   DEBUG <general>: CWSDiscoveryListenerUDP::ParseBuffer - Actionable message
sh: nmblookup: command not found
2021-09-28 13:21:12.216826+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.216 T:9393595   ERROR <general>: Unable to lookup host: 'FuzzNAS'
2021-09-28 13:21:12.217212+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.217 T:9393595   DEBUG <general>: CurlFile::Open - <http://FuzzNAS:5357/55b91b1c-e3d2-4b1b-a120-2a6eae5e8652>
2021-09-28 13:21:12.217503+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.217 T:9393595   DEBUG <general>: easy_acquire - Created session to http://FuzzNAS
2021-09-28 13:21:12.221112+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.221 T:9393595   ERROR <general>: CCurlFile::CReadState::FillBuffer - (0x12f83d820) Failed: Couldn't resolve host name(6)
2021-09-28 13:21:12.221315+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.221 T:9393595   ERROR <general>: CCurlFile::Open - <http://FuzzNAS:5357/55b91b1c-e3d2-4b1b-a120-2a6eae5e8652> Failed with code 0:
2021-09-28 13:21:12.221559+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.221 T:9393595   DEBUG <general>: CWSDiscoveryListenerUDP::UnicastGet - Could not fetch from http://FuzzNAS:5357/55b91b1c-e3d2-4b1b-a120-2a6eae5e8652
2021-09-28 13:21:12.221887+1000 kodi.bin[44672:9393595] 2021-09-28 13:21:12.221 T:9393595   DEBUG <general>: CWSDiscoveryUtils::printstruct - message contents
                                                   	Action: http://schemas.xmlsoap.org/ws/2005/04/discovery/ProbeMatches
                                                   	MsgID: urn:uuid:1a8774d1-74f6-4070-884e-f97459458913
                                                   	Address: urn:uuid:59c48269-8f1f-4051-9f13-529974bdaa69
                                                   	Types: wsdp:Device pub:Computer
                                                   	XAddrs: http://FuzzNAS:5357/55b91b1c-e3d2-4b1b-a120-2a6eae5e8652
                                                   	Computer:
```

garbear commented on xbmc/platform/posix/filesystem/SMBWSDiscoveryListener.cpp:
```diff
@@ -119,6 +120,33 @@ static const std::array<std::pair<std::string, std::string>, 1> address_tag{
+    "xmlns:wsa=\"http://schemas.xmlsoap.org/ws/2004/08/addressing\" "
+    "xmlns:wsd=\"http://schemas.xmlsoap.org/ws/2005/04/discovery\" "
+    "xmlns:wsdp=\"http://schemas.xmlsoap.org/ws/2006/02/devprof\" "
+    "xmlns:wsx=\"http://schemas.xmlsoap.org/ws/2004/09/mex\"> "
```
I always appreciate when line lists are organized alphabetically

mglae replied:
I don't get it. "xmlns:pnpx" ... "xmlns:wsx" are sorted.

garbear replied:
I'm saying that I appreciate that you sorted

garbear commented on xbmc/ServiceManager.cpp:
```diff
@@ -86,6 +90,9 @@ void CServiceManager::DeinitTesting()
   m_addonMgr.reset();
   m_databaseManager.reset();
+#if !defined(TARGET_WINDOWS) && defined(HAS_FILESYSTEM_SMB)
+  m_WSDiscovery.reset();
```
Can you remove the ifguards? `m_WSDiscovery.reset();` will do the correct thing always.

mglae replied:
Will do tomorrow - after Jenkins has completed his build.

mglae replied:
Done.

mglae replied:
Have to revert the change. m_WSDiscovery does not exist on all platforms.

mglae commented:
@fuzzard Thanks for testing.

Unicast get request can only be perforemed with a known IP address. Fixed, name resolution will not be triggered any more.

For the no servers in Add Video please provide a complete log.

fuzzard commented:
I'm getting a crash on Android with this build. 
I'll jump on computer (OSX) later today and see if it happens there. Nothing on the log on the android device, except it crashes to home screen immediately after the wsdiscovery service is started. I'll try a nightly (will have a new one run today in a few hours) and just confirm master isnt broke due to something else

fuzzard commented:
OSX doesnt crash at least. Still gives nothing when opening the "Windows Network (SMB)" in add files.

https://paste.kodi.tv/toyuzayoyo.kodi

Nothing i can see in the log will be of any use. Dont have time to debug it right now.

fuzzard commented:
Ok, when the server uses a name (FuzzNAS in my case), xaddrs_ip and computer are never set. Which bails out when you get to https://github.com/xbmc/xbmc/blob/7f5d0ab9e196804949415ab36a39da9d8418282d/xbmc/platform/posix/filesystem/SMBWSDiscovery.cpp#L83-L84

Hence no entries are added. Guess you are missing handling a non ip based value or something?

mglae commented:
@fuzzard We likely did have different user experience what was displayed in the server list, You saw names from xaddr lines that maybe cannot be reached. In my case there have been only reachable IP addresses listed.

Now both cases should be visible.

From the experience with the test suite Android may crash when DNSNameCache::GetCached() is called before an instance of CWSDiscoveryPosix is created. I've added some quick and dirty test code to verify this assumption.

fuzzard commented:
Androids not crashing now at least. However neither android or osx are showing any available WSD instances.
Same issue as yesterday essentially. Your 2 new instances (Computer and xaddrs_ip) are empty if xaddrs is a hostname

Just confirming, in current master with the exact setup, the hostname shows in the Windows Network (SMB) list, and i can enter it and see all the shares of the device.

The WSD instance is from a synology NAS (uses some version of the common wsdd.py implementation).
You are correct the name doesnt resolve, but kodi master still works and samba shares are shown (and work).

I think this is still too aggressive maybe.

https://github.com/xbmc/xbmc/blob/9b22199eac96e7b4ccf2121a3a1006d0b338e583/xbmc/platform/posix/filesystem/SMBWSDiscovery.cpp#L86-L87

Anyway, i dont actually use WS-Discovery, so really doesnt bother me in the slightest if my setup isnt supported, but short of anyone else doing any testing, its unfortunately a regression from the existing implementation imo

mglae commented:
@fuzzard Sorry, I messed up the code when creating commits.

Only minimal changes to a) make it build again on all platforms (Jenkins has to confirm) and b) now allowing any host names beside IPV6 addresses.

Some additional cleanup is needed but it has to be work first.

fuzzard commented:
Jenkins build this please

mglae committed:
WS-Discovery: Use unicast Get to request computer name

See https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-pbsd/16aebf07-c58b-40ee-b9fb-ce0b01112c32

mglae committed:
DNSNameCache: lookup found ws-discovery computer names

mglae commented:
I've no idea why WIN-32 and WIN-64 were failing in testsuite. Everything below xbmc/platform/posix should not be build and changes in xbmc/ServiceManager.cpp and xbmc/network/DNSNameCache.cpp are protected with `#if !defined(TARGET_WINDOWS) && `.

Code is squashed, rebased to master and cleaned up. Changing ServiceManager.cpp is not required any more.

fuzzard commented on xbmc/network/DNSNameCache.cpp:
```diff
@@ -106,24 +112,39 @@ bool CDNSNameCache::Lookup(const std::string& strHostName, std::string& strIpAdd
+    // loop through all DNSname entries and see if strHostName is cached
+    for (int i = 0; i < (int)g_DNSCache.m_vecDNSNames.size(); ++i)
     {
-      strIpAddress = DNSname.m_strIpAddress;
-      return true;
+      CDNSName& DNSname = g_DNSCache.m_vecDNSNames[i];
+      if (DNSname.m_strHostName == strHostName)
+      {
+        strIpAddress = DNSname.m_strIpAddress;
+        return true;
+      }
     }
```
Whilst you are here, can i request you update this for loop to the following

```
    // loop through all DNSname entries and see if strHostName is cached
    for (auto DNSname : g_DNSCache.m_vecDNSNames)
    {
      if (DNSname.m_strHostName == strHostName)
      {
        strIpAddress = DNSname.m_strIpAddress;
        return true;
      }
    }
```

We are slowly trying to modernise things that get touched in PR's

fuzzard replied:
I noticed today that i stuffed up the above diff. Really should be 

```
    for (const auto& DNSname : g_DNSCache.m_vecDNSNames)
```

I can do a followup fix to save you having to redo it again

mglae replied:
Changed.

fuzzard replied:
Cheers, thanks for that.

fuzzard reviewed (approved):
Tested working ok on my setup for osx/android clients.
Thanks for sticking it out with my setup/testing. If you wouldnt mind just touching up that for loop, that would be perfect.

Windows build failures can be disregarded at the moment. Other issues with jenkins. The ios/tvos build failures are also from another PR not related to this.

mglae commented:
@fuzzard Loop is simplified.

Thanks again for testing.

garbear commented:
I had to do `sudo apt install samba-common-bin` for the error disappear. Should this be documented?

mglae commented:
@garbear Do you think about documentation in readme or by adding log messages?

Can you post a log with enabled WS-Discovery component logging. I've only seen @fuzzard 's and my configuration til now.

I'm wondering if adding a mDNS .local domain lookup for single host names can improve the behavior.

fuzzard commented:
Lets just throw this in. See what shakes out.
